pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -119,6 +124,103 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, push *mo
 	return resources, model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cacheStats.hits, cacheStats.hits+cacheStats.miss)}
 }
 
+// BuildDeltaClusters will generate the deltas (add and delete) for a given proxy. Currently, only service changes are reflected with deltas.
+// Otherwise, we fall back onto generating everything.
+func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, push *model.PushContext, updates *model.PushRequest) ([]*discovery.Resource, []*discovery.Resource, model.XdsLogDetails) {
+	clusters := make([]*cluster.Cluster, 0)
+	resources := model.Resources{}
+	removedResources := model.Resources{}
+	services := make([]*model.Service, 0)
+	deletedServices := make([]*model.Service, 0)
+	envoyFilterPatches := push.EnvoyFilters(proxy)
+	cb := NewClusterBuilder(proxy, push, configgen.Cache)
+	instances := proxy.ServiceInstances
+	// delta when only services are modified and we attempt deltas
+	// we can progressively optimize this
+	delta := updates != nil && allConfigKeysOfType(updates.ConfigsUpdated, gvk.Service)
+	// if we can't use delta, fall back to generate all
+	if !delta || updates == nil || len(updates.ConfigsUpdated) == 0 || updates.Full {
+		cl, lg := configgen.BuildClusters(proxy, push)
+		return cl, make([]*discovery.Resource, 0), lg
+	}
+	configs := model.ConfigsOfKind(updates.ConfigsUpdated, gvk.Service)
+	for s := range configs {
+		// get service that changed
+		service := proxy.SidecarScope.Service(host.Name(s.Name))
+		// SidecarScope.Service will return nil if the proxy doesn't care about the service.
+		if service == nil {
+			continue
+		}
+		// these are the services we need to push
+		if updates.UpdateTypes[s] == model.EventUpdate || updates.UpdateTypes[s] == model.EventAdd {
+			services = append(services, service)
+		}
+		// we also need to update proxies of deleted config for delta
+		if updates.UpdateTypes[s] == model.EventDelete {
+			deletedServices = append(deletedServices, service)
+		}
+	}
+	clusterCacheStats := cacheStats{}
+	switch proxy.Type {
+	case model.SidecarProxy:
```

Comment:
ill have an internal method that has most of this logic to avoid duplicates between `BuildClusters` and `BuildDeltaClusters`

Reply:
maybe as a followup, the below code should be reused, otherwise it will be hard to maintain

Reply:
Can you please leave a TODO to refactor this to a common method?